composer/datasets/in_context_learning_evaluation.py

```diff
@@ -70,6 +81,155 @@ def _get_fewshot_sample_idxs(dataset_size, num_fewshot, sample_idx):
     return fewshot_idxs
 
 
+class InContextLearningQATaskDataset(Dataset):
+    """A dataset that construct batches for in-context learning question answering evaluation
+
+    The input format is expected to be a jsonl file with the following fields:
+    - context: the question
+    - answer: the preferred answer to the question
+    - aliases: a list of aliases for the answer
+
+    Args:
+        dataset_uri (str): Either a local path, or a remote path beginning with ``s3://``, or another backend
+            supported by :meth:`composer.utils.maybe_create_object_store_from_uri`. Dataset must consist of rows of JSON data points with "context",
+            "answer", and "aliases". See tests/datasets/local_data/triviaqa_small.jsonl.
+        tokenizer (Union[transformers.PreTrainedTokenizer, transformers.PreTrainedTokenizerFast]): The tokenizer used to map between strings and token ids
+        batch_size (int): Size of a batch used for eval
+        max_seq_len (int): The maximum sequence length supported by the model
+        pad_tok_id (int): The special token reserved for padding batches
+        num_fewshot (int): The number of complete fewshot examples to prepend before each test example
+        prompt_string (str): Prompt string to put once before all fewshot examples/test examples (e.g. 'translate english to french')
+        example_delimiter (str): Separator that goes between individual (context, answer) pairs (e.g. '\n')
+        continuation_delimiter: (str): Separator that goes between context and answer in each example (e.g. '\nA: ')
+        destination_path (str): Temporary path to store downloaded datasets
+        question_prelimiter (str): String to put before each question (e.g. 'Q: ')
+    """
+
+    def __init__(self, dataset_uri: str, tokenizer: Union[transformers.PreTrainedTokenizer,
+                                                          transformers.PreTrainedTokenizerFast], max_seq_len: int,
+                 pad_tok_id: int, num_fewshot: int, prompt_string: str, example_delimiter: str,
+                 continuation_delimiter: str, destination_path: str, question_prelimiter: str):
+        try:
+            from datasets import load_dataset  # pyright: ignore [reportGeneralTypeIssues]
+        except ImportError as e:
+            raise MissingConditionalImportError(extra_deps_group='nlp',
+                                                conda_package='datasets',
+                                                conda_channel='conda-forge') from e
+        with dist.local_rank_zero_download_and_wait(destination_path):
+            if dist.get_local_rank() == 0:
+                get_file(dataset_uri, destination_path, overwrite=True)
```

Comment:
Currently, we use `dist.local_rank_zero_download_and_wait` to avoid dist barrier. However, if the download fails, this results in a deadlock -> NCCL timeout which is pretty nasty. What if we instead added a context that runs only local rank 0 and then uses a signal file (ensuring the path to signal file exists or use `/tmp/some_signal_file`)? This way we avoid nasty deadlocks with download failures

Reply:
I'm pretty sure if the file doesn't exist it crashes, but will double check.

Reply:
confirmed that it just fails cleanly if the file does not exist.

Reply:
adding the context you suggested is a good general idea, but going to skip for this pr